### DISCUSSION TITLE
External scanner behaves different

### DISCUSSION THREAD

**Discussion by mrdgo** [tree-sitter/tree-sitter#3656] [Q&A] [answered]
### Problem

# Background

I want to implement an external scanner for [tree-sitter-nu](https://github.com/nushell/tree-sitter-nu).

I want to detect three types of nodes: `raw_string_begin`, `raw_string_content`, `raw_string_end`.

The language wants these weird strings, like `r#' hello world, use single quotes ' freely'#`.

It allows to use as many `#`'s as you wish, and only a balanced number will terminate the string:

### valid
- `r#' hi '#`
- `r#####' hi '###, bye'#####`

### invalid
- `r#'hi'#'#`
- `r###'hi'####'###`

# Problem

`tree-sitter parse` and `tree-sitter test` parse the code differently.

# Solutions tried
- I read about the arbitrary calls to serialize and deserialize, so these are proof (except if you find  a bug?)
- Fixed a few bugs that kept the scanner from terminating on arbitrary input

# Link to my repo
- [my current fork](https://github.com/mrdgo/tree-sitter-nu)
- relevant branch: `raw-strings`

### Steps to reproduce

```sh
gh repo clone mrdgo/tree-sitter-nu
cd tree-sitter-nu
git checkout raw-strings
tree-sitter test
```
Now you can see an error: tree-sitter detected a command and a comment.

But on the other hand, if you create a file `raw-string.nu` containing exactly `r#'string'#`
```sh
tree-sitter parse raw-string.nu
```
You will see that it parses the raw string correctly.

### Expected behavior

I want the tests in `tests/corpus/decl/raw-strings.nu` to pass.

### Tree-sitter version (tree-sitter --version)

tree-sitter 0.23.0 (12fb31826b8469cc7b9788e72bceee5af1cf0977)

### Operating system/version

Arch Linux

**Comment by ObserverOfTime** [comment]
See the [C++ scanner](https://github.com/tree-sitter/tree-sitter-cpp/blob/master/src/scanner.c).

**Reply by mrdgo** [reply]
What part of it? It generally does the same thing that I do. Do you have a detail in mind that is different?

**Comment by amaanq** [comment, marked as answer]
First, it'd be nice if your branch was actually a viable reproducer, but you had a logic bug regarding operator precedence [here](https://github.com/mrdgo/tree-sitter-nu/blob/6b7e5ce3f58af08c4372e6d9783a1d4c722f95f6/src/scanner.c#L156) (the or condition needs to be wrapped in parenthesis).

Second, your issue is that the test itself contains a newline before the string. You can easily parse the test with `ts p -n <test num>` (in this case, 67), and you can observe the unexpected parse tree. The newline before the string is the issue, how you want to handle that is up to you but one way is to skip spaces before lexing the raw string begin.

**Reply by mrdgo** [reply]
Wow thank you! I'll look into this, later. I appreciate your time. 

And yes, I committed some other things in the meanwhile, I am sorry for that.

**Reply by mrdgo** [reply]
Yeah, I somehow confused `||` and `&&`.

That actually fixed it. Thank you so much again :pray:
